src/pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -583,6 +583,8 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
         """
         if self.dirpath is not None:
             # short circuit if dirpath was passed to ModelCheckpoint
+            if broadcast:
+                self.dirpath = trainer.strategy.broadcast(self.dirpath)
```

Comment:
This is too dangerous for my taste. It can lead to unwanted hangs when dirpath is None on one rank and not None on another. Even if it's not the case today, it's annoying to debug if you run into it. Our best practice in Lightning is to never run collectives under a conditional unless you can guarantee that every rank enters it. 

I suggest to keep the broadcasting outside of `__resolve_ckpt_dir`, i.e., wherever it is needed, call

```
dir = __resolve_ckpt_dir()
dir = broadcast(dir)
```
and otherwise not. 

What do you think?

Reply:
That absolutely makes sense. Will make the required changes

Reply:
Done. Thank you!